### DISCUSSION TITLE
Organization of the finals

### DISCUSSION THREAD

**Discussion by omichel** [cyberbotics/wrestling#63] [Announcements]
I believe it is important to give some precision on how the finals will be held during the upcoming ICRA 2023 conference in London. It turns out that some participant, namely SugarSkull, is running many games in a raw (e.g., 35 games recently) with possibly different parameters or algorithms until they win. Although this strategy seems to be pretty efficient during the qualification stage (e.g., now), it won't work any more during the finals. Indeed, to prepare the finals, we will checkout the controller of each participant on May 23rd between 2:00pm and 4:00pm (BST). We will use only this version of the participant controller programs to run the final games. This includes the 1/16, 1/8, 1/4, 1/2 and third place and final game. So, you'd better have a very robust robot controller ready by then, as it will be the only version used throughout the finals.

I will add this explanation to the [README.md](https://github.com/cyberbotics/wrestling#icra-2023-humanoid-robot-wrestling-competition) file, so that it is very clear for newcomers as well.

**Comment by omichel** [comment]
See the [update](https://github.com/cyberbotics/wrestling/commit/14a04aedc04381244b7ca80fdd04e2b3591f959b) of the README.md file.

**Comment by JustLuckyRobotics** [comment]
Would there be consideration to run each finals fight best of 3? RNG plays heavily into these fights and if the most robust controller is desired, best of 3 would highlight this more effectively. Most fights would likely only be ran twice to confirm the outcome, those that need to do the full 3 could also give entertaining clips as they duke it out. 

Having the referee call a simulation reset on the decision of the fight would immediately start the next fight without the need re do the docker building etc. to save time potentially?

if this is already the intention and i have missed it in the documentation i apologise. I also understand if this would cause the finals to take too long so might not be considered.

**Reply by omichel** [reply]
That is a very interesting proposal indeed.

PRO:
- reduce the impact of bad luck in the games.
- make some games more entertaining when there is some suspense about the outcome.

CON:
- the games could last up to 9 minutes (instead of 3).

We are pretty positive about it. So, if there is no motivated objection about this on this discussion thread, we will modify the rules to implement this two-winning-sets rule for the finals and document it in the README.md file.

**Reply by zicklag** [reply]
I think that makes a lot of sense. It would be lame if it seemed like the final ended up being lost by one competant bot which got unlucky.

**Comment by omichel** [comment]
I just updated the main [README.md](https://github.com/cyberbotics/wrestling/tree/main#readme) to provide more details about the organization of the finals.

I also prepared a [result](https://github.com/cyberbotics/wrestling/blob/main/result.md) page that we will keep up-to-date when the finals start.

The deadline of [May 23rd](https://github.com/cyberbotics/wrestling/blob/main/README.md#selection-for-the-finals) is approaching!

Let me know if you have any question about it.

**Comment by omichel** [comment]
Today at [2pm BST](https://dateful.com/convert/british-summer-time-bst?t=14&d=2023-05-23) sharp, I will stop the entrance of participants in the evaluation queue. I will wait until the queue is empty (e.g., all evaluations are complete) and will freeze the ranking that will determine the 32 teams selected for the finals. Then, I will download the controller repositories of the 32 selected teams. Please refrain from making any modification to your controller repository from [2pm BST](https://dateful.com/convert/british-summer-time-bst?t=14&d=2023-05-23) today or do it at your own risk. I will post a message here when the download of controller repositories is complete. Only these downloaded controllers will be used for the finals.

**Comment by omichel** [comment]
The evaluation queue is now closed. However, there are currently 9 participants in the queue. I will wait until it finishes to determine the final ranking of the 32 selected teams. I am starting now to download the controllers of the top 32 participants.

Will [HeavyRotation](https://github.com/YuriCat) enter the top 32? That would be fantastic...

**Reply by omichel** [reply]
You can now sit back and relax 😉

**Reply by MJavadZallaghi** [reply]
This last hours was really stressful. :)

**Comment by omichel** [comment]
We now have the ranking for the finals! :tada:

**Comment by omichel** [comment]
I just completed the download of the repositories. The leaderboard is all set for the finals!

Thank you all for your great effort and fair-play in participating in this competition. It was a lot of fun for me to organize and I am looking forward to the finals now!

Alea iacta est!

**Reply by MJavadZallaghi** [reply]
Thank you for providing a lot of learning opportunities. I'm so happy to be in this awesome competition, whatever the result would be.

**Reply by AndrejOrsula** [reply]
Thank you for organizing this competition and putting in so much effort! It has been a fantastic opportunity to explore Webots, especially with the easy-to-use infrastructure that you provided.

Best of luck to all the participants in the finals!

**Comment by omichel** [comment]
I just updated the [1/16 finals table](https://github.com/cyberbotics/wrestling/blob/main/result.md#116-finals).

**Reply by amirhnourian** [reply]
Hi, dear Omichel:
Unfortunately, I was not aware that we can not update our controller after this time and this is our controller for the final game, can you give us a chance to push our final controller for the final game?

**Reply by amirhnourian** [reply]
I have just miss about 10 minutes :DD

**Reply by omichel** [reply]
I am very sorry but the download of snapshots is now over and I cannot accept further changes. This was clearly stated in this discussion thread and in the [main page](https://github.com/cyberbotics/wrestling/tree/main#selection-for-the-finals). Accepting your request wouldn't be fair to other participants.

**Reply by amirhnourian** [reply]
Thank you for the opportunity! It was a great experience!
As in my last push, I changed my JSON file and changed my account type to single. Please kindly if you are going to give us any certificate use the last JSON file. 
I appreciate all your efforts and great software!

**Reply by MehdiZouitine** [reply]
Hello thank's you for this great competition !

So as I understand, we cannot change our controllers for the final phase ?

**Reply by omichel** [reply]
@amirhnourian: Yes, I will use your latest information from your JSON file for the certificate if any. We plan to provide official ICRA certificates only to the top 3.

**Reply by omichel** [reply]
@MehdiZouitine: you are right, controllers cannot be changed any more. You just have to cross your fingers now... 😉.

**Reply by amirhnourian** [reply]
It's unfortunate so are we going to get any certificate? I did not check the time given for the final controller and thought we have one week to work on our controllers even after the 32 Team selection! 
Thank you for your time and prompt response

**Reply by omichel** [reply]
Cyberbotics will make certificates for all 32 finalists. However, only the 3 first ones will have the official ICRA signature and sealing stamp. Others will have Cyberbotics signature and sealing stamp.

**Reply by hengyue-dunhuang** [reply]
Is that the digital certificate ? Can I just update the participant json cause I don't mention my team members including me?🙏

**Reply by omichel** [reply]
Sure, please let me know when done.

**Reply by hengyue-dunhuang** [reply]
I have already updated my participant.json. Thank you very much!

**Reply by hengyue-dunhuang** [reply]
Oh , the team name is Wrestler Zhang, which may be helpful. Thanks!

**Reply by omichel** [reply]
Thank you.
